circuit_knitting/cutting/cut_finding/cco_utils.py

```diff
@@ -138,23 +138,25 @@ def greedy_best_first_search(
         Callable, search_space_funcs.next_state_func
     )
 
-    if search_space_funcs.goal_state_func(state, *args):
-        return state
-
-    best = min(
-        [
-            (search_space_funcs.cost_func(next_state, *args), k, next_state)
-            for k, next_state in enumerate(
-                search_space_funcs.next_state_func(state, *args)
-            )
-        ],
-        default=(None, None, None),
-    )
-
-    if best[-1] is None:  # pragma: no cover
-        # This covers a rare edge case.
-        # We have so far found no circuit that triggers it.
-        # Excluding from test coverage for now.
-        return None
-
-    return greedy_best_first_search(best[-1], search_space_funcs, *args)
+    while True:
```

Comment:
Do you have any thoughts about cases where this could run infinitely? What are the cases where `142` wouldn't evaluate to `True`?

Just a suggestion, but I like to bring return statements out of the loop if I can for clarity. I also like to include the condition in the while statement if I can to make it a little more clear when this thing should stop

```
>>> while not search_space_funcs.goal_state_func(...):
>>>     best = min(
>>>     ...
>>>     state = best[-1]
>>>     continue
>>> return state
```

Just a suggestion, it's up to you

Reply:
>Do you have any thoughts about cases where this could run infinitely? What are the cases where #142 wouldn't evaluate to True ever or for a long time?

So the way it works is that a circuit is represented as a list of gates and each gate in this list is ''visited,'' one after the other, whereby a decision on whether or not to cut and how to cut (wire vs gate) is made. Once all the gates have been visited (in the sense described above), the state obtained is a goal state. That is basically when the loop terminates. Therefore, the loop will run for longer when there is a large number of gates in the circuit.

Reply:
OK, so this is really an O(N) loop under the hood, where N is the number of gates in the circuit. There isn't a condition where it would run infinitely or for an unknown amount of time that we can think of?

Reply:
I think that's right, yes.

Reply:
>Just a suggestion, but I like to bring return statements out of the loop if I can for clarity. I also like to include the condition in the while statement if I can to make it a little more clear when this thing should stop.

 I think this sounds good. Will work on this.

Reply:
Updated: [dc43c78](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/pull/537/commits/dc43c789526fa975fb75819f06e9a1ce70dce6aa)